pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/QualifiedName.java

```diff
@@ -97,11 +110,47 @@ public static QualifiedName makeOuterClassOf(ASTClassOrInterfaceDeclaration node
 
 
     // Might be useful with type resolution
-    public static QualifiedName parseCanonicalName(String canon) {
+    public static QualifiedName makeClassOf(Class<?> clazz) {
         throw new UnsupportedOperationException();
     }
 
-    /** Returns a normalized method name (not Java-canonical!) */
+    /**
+     * Parses a qualified name given in the format defined for this implementation. The format
+     * is described as the following regex pattern :
+     *
+     * <p>{@code ((\w+\.)+|\.)((\w+)(\$\w+)*)(#(\w+)\(((\w+)(, \w+)*)?\))?}
```

Comment:
Since you only need 3 capturing groups, you could define the other groups as a  [non-capturing group](https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html): `(?:...)` - With the downside, that the regex is even worse readable....

It should be for this regex:

    ((?:\w+\.)+|\.)((?:\w+)(?:\$\w+)*)(#(?:\w+)\((?:(?:\w+)(?:, \w+)*)?\))

For the documentation a simple example is probably easier to understand:

    com.company.MyClass$Nested#myMethod(String, int)

I would then just refer the the regex (@see #FORMAT)